stan/math/prim/fun/add.hpp

```diff
@@ -11,56 +12,49 @@ namespace math {
  * Return the sum of the specified matrices.  The two matrices
  * must have the same dimensions.
  *
- * @tparam T1 type of elements in the first matrix
- * @tparam T2 type of elements in the second matrix
- * @tparam R number of rows, can be Eigen::Dynamic
- * @tparam C number of columns, can be Eigen::Dynamic
+ * @tparam T1 type of the first matrix or expression
+ * @tparam T2 type of the second matrix or expression
  *
- * @param m1 First matrix.
- * @param m2 Second matrix.
+ * @param m1 First matrix or expression.
+ * @param m2 Second matrix or expression.
  * @return Sum of the matrices.
  * @throw std::invalid_argument if m1 and m2 do not have the same
  * dimensions.
  */
-template <typename T1, typename T2, int R, int C>
-inline Eigen::Matrix<return_type_t<T1, T2>, R, C> add(
-    const Eigen::Matrix<T1, R, C>& m1, const Eigen::Matrix<T2, R, C>& m2) {
+template <typename T1, typename T2, typename = require_all_eigen_t<T1, T2>>
+inline auto add(const T1& m1, const T2& m2) {
   check_matching_dims("add", "m1", m1, "m2", m2);
-  return m1 + m2;
+  return (m1 + m2).eval();
```

Comment:
Since there is no `.array()` in this expression and we assume all input are matrices the output will also be matrix. If someone passes in two arrays I would expect this to return the same type - an array.

Reply:
We previously returned a matrix so I think it would be good to return a matrix now. When we start unwinding these `evals()` we can let this return an array

Reply:
Previously this function did not accept arrays, so there is no previous behavour we would have to be consistent with.

Reply:
The problem here is that if we miss another function that ends up returning an array, then pump that function into multiply, then pump multiply into a function that only accepts a matrix this is gonna throw when we try to compile. Until the rest of stan math can accept non-matrix eigen expressions being safe feels much better here

Reply:
Not compiling or throwing if a user makes a bug seems fine to me. Anyway using `Eigen::Array` is not possible from stan language so potential for using Arrays is internal to Stan Math. I think we should follow the convention that functions return same type as they got in, if that makes sense. Std math functions do this, Eigen does it ...

Also adding `.matrix()` would not work for matrices, as they do not define this member function.  Instead we would have to write separate set of overloads for `Eigen::Array`s, which we don't actually need.